core/src/apps/webauthn/credential.py

```diff
@@ -204,9 +231,53 @@ def private_key(self) -> bytes:
         path = [HARDENED | 10022, HARDENED | int.from_bytes(self.id[:4], "big")] + [
             HARDENED | i for i in ustruct.unpack(">4L", self.id[-16:])
         ]
-        node = seed.derive_node_without_passphrase(path, "nist256p1")
+        node = seed.derive_node_without_passphrase(path, _CURVE_NAME[self.curve])
         return node.private_key()
 
+    def public_key(self) -> bytes:
+        if self.curve == common.COSE_CURVE_P256:
+            pubkey = nist256p1.publickey(self.private_key(), False)
+            return cbor.encode(
+                {
+                    common.COSE_KEY_ALG: self.algorithm,
+                    common.COSE_KEY_KTY: common.COSE_KEYTYPE_EC2,
+                    common.COSE_KEY_CRV: self.curve,
+                    common.COSE_KEY_X: pubkey[1:33],
+                    common.COSE_KEY_Y: pubkey[33:],
+                }
+            )
+        elif self.curve == common.COSE_CURVE_ED25519:
+            pubkey = ed25519.publickey(self.private_key())
+            return cbor.encode(
+                {
+                    common.COSE_KEY_ALG: self.algorithm,
+                    common.COSE_KEY_KTY: common.COSE_KEYTYPE_OKP,
+                    common.COSE_KEY_CRV: self.curve,
+                    common.COSE_KEY_X: pubkey,
+                }
+            )
+        raise TypeError
+
+    def sign(self, data: Iterable[bytes]) -> bytes:
+        if (self.algorithm, self.curve) == (
+            common.COSE_ALG_ES256,
+            common.COSE_CURVE_P256,
+        ):
+            dig = hashlib.sha256()
+            for segment in data:
+                dig.update(segment)
+            sig = nist256p1.sign(self.private_key(), dig.digest(), False)
+            return der.encode_seq((sig[1:33], sig[33:]))
+        elif (self.algorithm, self.curve) == (
+            common.COSE_ALG_EDDSA,
+            common.COSE_CURVE_ED25519,
+        ):
+            return ed25519.sign(
+                self.private_key(), b"".join(bytes(segment) for segment in data)
```

Comment:
Not necessarily, because MicroPython is totally messed up when it comes to handling arrays of bytes and mainly because the people around mypy decided to make things even more complicated: https://github.com/python/typing/issues/552.

I tried to rectify the bytes/bytearray typing around uctypes as best I could. Although I am not entirely satisfied with it, I think it's in accordance with what we discussed.

Reply:
Got it.